### ISSUE TITLE
SSM connect plugin uses SSM Session Manager and opens/closes multiple connections per single ansible execution. Consider Run Command as alternative.

### ISSUE THREAD

**Issue by mdaehnert** [ansible-collections/community.aws#1148]
### Summary

When SSM connect plugin is leveraged to call EC2 machines, Ansible opens and closes multiple sessions per request. (i.e. one session for echo, one for transferring the execution file, one for executing it, ...). This behavior leads to ~3-6 SSM sessions opened and closed for a single ```Get-Location```/```pwd```.

Its documentation isn't quite clear, but SSM Session Manager is intended to be used by interactive remote shells but not for highly frequent automation tasks (sessions opened and closed).

One of my customers experienced unresponsive SSM agents on EC2 (mostly tested with Windows hosts). They do regular security checks via Ansible+SSM plugin (multiple times per hour) which causes their SSM agents to die after some days. A penetration test could even kill the agent within hours.

A more resilient approach to automate tasks on EC2 is via SSM Run Command. Its an asynchronous way to execute automation commands at scale on EC2 machines.

I used the existing SSM connect plugin and changed about 10% of its source code to work with SSM Run Command. I would like to optimize the source code and create a new PR to this repository. How about Run Command as second file/option so that Ansible users can choose between Session Manager and Run Command usage.

Thanks for your feedback.

Pro:
* Run Command is intended for high frequency of automation tasks

Contra:
* Run Command runs asynchronously and takes about 3 times the time like Session Manager

### Issue Type

Bug Report

### Component Name

ssm connect plugin

### Ansible Version

unrelated

### Collection Versions

unrelated

### AWS SDK versions

unrelated

### Configuration

unrelated

### OS / Environment

unrelated

### Steps to Reproduce

Execute ssm exeuction multiple times in a row.

### Expected Results

SSM agent on EC2 should work as expected even with hundreds of executions per hour over a long period of weeks/months.

### Actual Results

Systems Manager Agent on Windows/Linux machines becomes unresponsive. There are multiple issues shown on Windows machines: Sometimes file handles won't be released by SSM agent and the agent stops responding, sometimes the agent stops responding without proper information on logs.

### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

**Comment by ansibullbot** [comment]
Files identified in the description:
None

If these files are inaccurate, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by mdavis-xyz** [comment]
I don't understand. What is the benefit of run command over running commands through a session?


> Its an asynchronous way

Isn't that a bad thing, not a good thing, in this context? (Since Ansible tasks are synchronous by default.)

**Comment by mdaehnert** [comment]
The goal of this is to mitigate the failing sessions with the initial approach. Yes, the async approach via run command is a technical pushback, but it mitigates failing sessions at high load. On top: run command is intended for technical automations at scale. Based on AWS docimentation, session manager is for human interactions with a few sessions in parallel only.

**Comment by mdavis-xyz** [comment]
What's the problem with running SSM sessions a lot?

> causes their SSM agents to die after some days

This sounds like a bug in the SSM agent. And it should be configured to autorestart.

How does run command mitigate this? run command uses the same agent, the same websocket etc.

In #1853 I'm trying to change the plugin to connection one SSM session per playbook run, instead of one per file transfer, which is several per task. Wouldn't that be even better than a run command connection per file transfer (which would still be several per task)?

**Comment by mdaehnert** [comment]
Back in 2022, we talked to the AWS team who was responsible for the agent as well. They argued that the frequency of sessions was way too high. The issue was, that the agent got unresponsive with sometimes a lot of handles open (Windows), sometimes without any information seen on task manager. Customer solved it by rebooting EC2s every x-days.

With Run Command the SSM server queues calls and will not overburden the agent by design. Session Manager will start a DDoS attack if requested. ;⁠-⁠)

Your fix might solve the issue already. I also intended to see if other customers face the issue, but you're the first one replying in a year. So go for it as it will improve the usage of SSM without having the async/performance implications of Run Command.

**Comment by mdavis-xyz** [comment]
If you do still want a run command option, it might be better to write a new connector from scratch.  There's a new [httpapi](https://docs.ansible.com/ansible/latest/plugins/httpapi.html) connection plugin that seems relatively new. It might be more idiomatic and elegant to use that instead of the existing `aws_ssm.py`. A lot of the existing `aws_ssm.py` code is about generating random UUIDs to print out before and after the actual command, to find the part of the stdout to extract. But for run command that's not required.

**Comment by abikouo** [comment]
@mdaehnert Runcommand is asynchronous, this can't solve the issue with this plugin.
I am closing this issue for now.
